sopel/bot.py

```diff
@@ -592,6 +592,58 @@ def register_urls(self, urls: Iterable) -> None:
                 except plugins.exceptions.PluginError as err:
                     LOGGER.error("Cannot register URL callback: %s", err)
 
+    def rate_limit_info(self, rule: plugin_rules.Rule, trigger: Trigger) -> Tuple[bool, Optional[str]]:
+        if trigger.admin or rule.is_unblockable():
+            return False, None
+
+        is_channel = trigger.sender and not trigger.sender.is_nick()
+        channel = trigger.sender if is_channel else 'private-message'
+
+        at_time = trigger.time
+
+        user_metrics = rule.get_user_metrics(trigger.nick)
+        channel_metrics = rule.get_channel_metrics(channel)
+        global_metrics = rule.get_global_metrics()
+
+        if user_metrics.is_limited(at_time - rule.user_rate_limit):
+            template = rule.user_rate_template
+            rate_limit_type = "user"
+            rate_limit = rule.user_rate_limit
+            metrics = user_metrics
+        elif is_channel and channel_metrics.is_limited(at_time - rule.channel_rate_limit):
+            template = rule.channel_rate_template
+            rate_limit_type = "channel"
+            rate_limit = rule.channel_rate_limit
+            metrics = channel_metrics
+        elif global_metrics.is_limited(at_time - rule.global_rate_limit):
+            template = rule.global_rate_template
+            rate_limit_type = "global"
+            rate_limit = rule.global_rate_limit
+            metrics = global_metrics
+        else:
+            return False, None
+
+        next_time = metrics.last_time + rate_limit
+        time_left = next_time - at_time
+
+        if template:
+            message = template.format(
+                nick=trigger.nick,
+                channel=channel,
```

Comment:
I'd ensure here that the `channel` key is `str` (goes with the previous issue of "channel should be `None`"):

```suggestion
                channel=channel or 'private message',
```